### PR TITLE
Introduce Export/Import (to file) Reading List

### LINKED ISSUE #61: Allow option to export bookmarks
Core should be in the libkiwix and available through JNI (see https://github.com/kiwix/kiwix-android/issues/822)

### PR DESCRIPTION
**Pull request by ShaopengLin**
Fix #61 

Changes:
- User can Export to or Import from a file.
- Added Menu Button to File section of Main Menu for Ex/Import Reading List.
- Added UI on Reading List Bar for Ex/Import.

### CODE REVIEW

**Comment by ShaopengLin** [comment]
@veloman-yunkan The readBookmarkFile simply appends all bookmarks to the library. This can create duplicate bookmark entries. Should we do the elimination on Kiwix-Desktop or libkiwix?

**Comment by kelson42** [comment]
@ShaopengLin Thank you for the PR, I will give a try today.

**Review by kelson42 — APPROVED** [review]
LGTM but please use OS "Download" directory instead of current directory for the default filepicker location
@veloman-yunkan ready for code review

**Comment by kelson42** [comment]
@veloman-yunkan We need urgently a code review.

File: src/readinglistbar.cpp [hunk 1]
```diff
@@ -122,6 +124,20 @@ void ReadingListBar::onExport()
         app->showMessage(gt("export-reading-list-error"), gt("error-title"), QMessageBox::Information);
 }
 
+void ReadingListBar::onImport()
+{
+    auto app = KiwixApp::instance();
+    auto library = app->getLibrary();
+    QString fileName = QFileDialog::getOpenFileName(app->getMainWindow(),
+                                                    gt("save-file-as-window-title"),
```
Comment (veloman-yunkan) [inline src/readinglistbar.cpp:141]: "Save file as..." is a misleading window title when opening a file

File: src/readinglistbar.cpp [hunk 2]
```diff
@@ -6,6 +6,9 @@
 
 #include <QListWidgetItem>
 #include <QFileDialog>
+#include <QStandardPaths>
+
+const QString downloadDir = QStandardPaths::writableLocation(QStandardPaths::DownloadLocation);
```
Comment (veloman-yunkan) [inline src/readinglistbar.cpp:11]: Is the Downloads directory a good default location? Isn't Documents a better one?

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]

File: src/readinglistbar.cpp (hunk 2 above)
Comment (ShaopengLin) [reply src/readinglistbar.cpp:11]: @kelson42 I checked the chromium im/export and they use Document as the default location.

File: src/readinglistbar.cpp [hunk 3]
```diff
@@ -153,3 +153,12 @@ void ReadingListBar::openUrl(QListWidgetItem* item, bool newTab)
...
 }
+
+void PortMenu::showEvent(QShowEvent* event)
+{
+    Q_UNUSED(event);
+
+    QPoint p = this->pos();
+    QRect geo = b->geometry();
+    this->move(p.x()+geo.width() - 4, p.y() - b->iconSize().height());
+}
```
[the remark this answers is not in the document]
Comment (ShaopengLin) [reply src/readinglistbar.cpp:168]: I don't think its possible as the position seems to be hardcoded by Qt. Found no style exported to change this.

File: src/readinglistbar.cpp (hunk 2 above)
Comment (kelson42) [reply src/readinglistbar.cpp:11]: OK, then go for "Documents"

File: src/readinglistbar.cpp (hunk 1 above)
Comment (ShaopengLin) [reply src/readinglistbar.cpp]: Done

File: src/readinglistbar.cpp (hunk 2 above)
Comment (ShaopengLin) [reply src/readinglistbar.cpp]: Done

**Comment by veloman-yunkan** [comment]
> @veloman-yunkan The readBookmarkFile simply appends all bookmarks to the library. This can create duplicate bookmark entries. Should we do the elimination on Kiwix-Desktop or libkiwix?

@kelson42 Should we address the above issue in this PR?

File: src/readinglistbar.cpp [hunk 4]
```diff
@@ -122,6 +124,20 @@ void ReadingListBar::onExport()
         app->showMessage(gt("export-reading-list-error"), gt("error-title"), QMessageBox::Information);
 }
 
+void ReadingListBar::onImport()
+{
+    auto app = KiwixApp::instance();
+    auto library = app->getLibrary();
+    QString fileName = QFileDialog::getOpenFileName(app->getMainWindow(),
+                                                    gt("open-file"),
+                                                    QString(), "(*.xml)");
```
Comment (veloman-yunkan) [inline src/readinglistbar.cpp]: Please make the filter string more user-friendly, change it to `XML files (*.xml)`

File: resources/i18n/en.json [hunk 5]
```diff
@@ -173,5 +173,7 @@
     "file-not-found-text": "ZIM file doesn't exist or is not readable",
     "zim-id": "Zim Id",
     "zim-name": "Zim Name",
-    "zim-path": "Zim File Path"
+    "zim-path": "Zim File Path",
+    "export-reading-list": "Export reading list",
+    "export-reading-list-error": "An error has occured during export."
```
Comment (veloman-yunkan) [inline resources/i18n/en.json]: @kelson42 Are you fine with such an error message? I don't like that "reading list" is missing from the error text. Also no details about the error are provided, but addressing that one will require changes to `libkiwix`, and most likely that part won't be translated.

File: resources/i18n/en.json [hunk 6]
```diff
@@ -175,5 +175,7 @@
     "zim-name": "Zim Name",
     "zim-path": "Zim File Path",
     "export-reading-list": "Export reading list",
-    "export-reading-list-error": "An error has occured during export."
+    "export-reading-list-error": "An error has occured during export.",
+    "import-reading-list": "Import reading list",
+    "import-reading-list-error": "An error has occured during import."
```
Comment (veloman-yunkan) [inline resources/i18n/en.json:180]: @kelson42 Are you fine with such an error message? I don't like that "reading list" is missing from the error text. Also no details about the error are provided, but addressing that one will require changes to `libkiwix`, and most likely that part won't be translated.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]

File: src/readinglistbar.cpp (hunk 4 above)
Comment (ShaopengLin) [reply src/readinglistbar.cpp]: @veloman-yunkan do we need to translate "XML files" or does Qt does it automatically?

Comment (veloman-yunkan) [reply src/readinglistbar.cpp]: That's a good question. But I guess that the answer to it is negative. So maybe it is better to keep the filter language neutral.

File: resources/i18n/en.json (hunk 5 above)
Comment (ShaopengLin) [reply resources/i18n/en.json]: Changed the message as requested and waiting for confirmation.

File: resources/i18n/en.json (hunk 6 above)
Comment (ShaopengLin) [reply resources/i18n/en.json]: Changed the message as requested and waiting for confirmation.

File: src/readinglistbar.cpp [hunk 7]
```diff
@@ -104,6 +108,20 @@ void ReadingListBar::onItemActivated(QListWidgetItem* item, Qt::MouseButtons but
     }
 }
 
+void ReadingListBar::onExport()
+{
+    auto app = KiwixApp::instance();
+    auto kiwixLibrary = app->getLibrary()->getKiwixLibrary();
+    QString fileName = QFileDialog::getSaveFileName(app->getMainWindow(),
+                                                    gt("save-file-as-window-title"),
+                                                    "readinglist.xml");
```
[the remark this answers is not in the document]
Comment (ShaopengLin) [reply src/readinglistbar.cpp]: Done

[the remark this answers is not in the document]
Comment (ShaopengLin) [reply src/readinglistbar.cpp]: Done

**Review by veloman-yunkan — COMMENTED** [review]
LGTM. Not approving yet, since there is [an unanswered question](https://github.com/kiwix/kiwix-desktop/pull/1163#issuecomment-2277470304).

File: resources/i18n/en.json (hunk 5 above)
Comment (kelson42) [reply resources/i18n/en.json]: @veloman-yunkan Yes, message should be "An error has occured during reading list export."

**Comment by kelson42** [comment]
> > @veloman-yunkan The readBookmarkFile simply appends all bookmarks to the library. This can create duplicate bookmark entries. Should we do the elimination on Kiwix-Desktop or libkiwix?
> 
> @kelson42 Should we address the above issue in this PR?

**Comment by kelson42** [comment]
> > @veloman-yunkan The readBookmarkFile simply appends all bookmarks to the library. This can create duplicate bookmark entries. Should we do the elimination on Kiwix-Desktop or libkiwix?
> 
> @kelson42 Should we address the above issue in this PR?

@veloman-yunkan Very good point, I have open a dedicated issue as this is not trivial to fix and we can leave IMHO with this weakness https://github.com/kiwix/kiwix-desktop/issues/1171

**Pull request merged**